src/perf/best-practices.md

```diff
@@ -0,0 +1,375 @@
+    animations where the animation doesn't affect the child subtree.
+    See the [`TransitionBuilder`][] pattern and
+    the [source code for `SlideTransition`][],
+    which uses this principle to avoid rebuilding its
+    descendents when animating.
+
+Use `const` constructors on widgets as much as possible,
+since they allow Flutter to short-circuit most of the rebuild work.
+To create reusable pieces of UIs,
+prefer using a [`StatelessWidget`][] rather than a function.
+
+For more information, check out:
+
+* [Performance considerations][],
+  part of the [`StatefulWidget`][] API doc
+* [Widgets vs helper methods][],
+  a video from the official Flutter YouTube 
+  channel that explains why widgets
+  (especially widgets with `const` constructors)
+  are more performant than functions.
+
+[Performance considerations]: {{site.api}}/flutter/widgets/StatefulWidget-class.html#performance-considerations
+[source code for `SlideTransition`]: {{site.repo.flutter}}/blob/master/packages/flutter/lib/src/widgets/transitions.dart#L168
+[`StatefulWidget`]: {{site.api}}/flutter/widgets/StatefulWidget-class.html
+[`StatelessWidget`]: {{site.api}}/flutter/widgets/StatelessWidget-class.html
+[`TransitionBuilder`]: {{site.api}}/flutter/widgets/TransitionBuilder.html
+[Widgets vs helper methods]: {{site.youtube-site}}/watch?v=IOyq-eTRhvo
+
+---
+
+## Use saveLayer() thoughtfully
+
+Some Flutter code uses `saveLayer()`, an expensive operation,
+to implement various visual effects in the UI.
+Even if your code doesn't explicitly call `saveLayer()`,
+other widgets or packages that you use might call it behind the scenes.
+Perhaps your app is calling `saveLayer()` more than necessary.
+
+{{site.alert.secondary}}
+  <h4>Why is savelayer expensive?</h4><a name="why-is-savelayer-expensive"></a>
+  Calling `saveLayer()` allocates an offscreen buffer.
+  Drawing content into the offscreen buffer might trigger
+  render target switches that are particularly slow in
+  older GPUs.
+{{site.alert.end}}
+
+How can you tell how much your app calls `saveLayer()`,
+either directly or indirectly?
+The `saveLayer()` method triggers
+an event on the [DevTools timeline][].
+[Pending: What else can we say here? For example,
+ how do you define _excessive_?]
+If the calls to `saveLayer()` are excessive,
+this can cause jank.
+
+[DevTools timeline]: {{site.url}}/development/tools/devtools/performance#timeline-events-chart
+
+You have several options to handle this situation:
+
+* If the calls are coming from _your_ code, can you
+  reduce or eliminate them? For example, perhaps your
+  UI overlaps two shapes, each having some transparency.
+  If they always overlap in the same amount, in the same
+  way, with the same transparency,
+  you can precalculate what this overlapped,
+  semi-transparent object looks like, cache it,
+  and use that instead of calling `saveLayer()`. 
+  This works with any static shape that you can
+  precalculate.
```

Comment:
@johnpryan @InMatrix this would be a great place to link to an example. Perhaps a new addition to [perf_dignosis_demo](https://github.com/InMatrix/perf_diagnosis_demo)?

Reply:
It might be useful to look at where and how `saveLayer` is used in pub packages using [survayor](https://github.com/pq/surveyor). It's possible that Flutter lacks some high-level APIs which pushed users towards `saveLayer`. Cc: @pq

Reply:
> It might be useful to look at where and how saveLayer is used in pub packages using [survayor](https://github.com/pq/surveyor).

That could be edifying for sure.  Happy to advise if anyone decides to look closer.